src/server/search/aggregator.cc

```diff
@@ -10,62 +10,89 @@ namespace dfly::aggregate {
+  }
+  return out;
+}
 
-    absl::flat_hash_set<std::string> fields_to_print;
-    fields_to_print.reserve(fields_.size() + reducers_.size());
+DocValues PackFields(ValuesList values, absl::Span<const std::string> fields) {
+  DCHECK_EQ(values.size(), fields.size());
+  DocValues out;
+  for (size_t i = 0; i < fields.size(); i++)
+    out[fields[i]] = std::move(values[i]);
+  return out;
+}
 
-    for (auto& field : fields_) {
-      fields_to_print.insert(std::move(field));
-    }
-    for (auto& reducer : reducers_) {
-      fields_to_print.insert(std::move(reducer.result_field));
-    }
+const Value kEmptyValue = Value{};
 
-    return {std::move(out), std::move(fields_to_print)};
+}  // namespace
+
+void Aggregator::DoGroup(absl::Span<const std::string> fields, absl::Span<const Reducer> reducers) {
+  // Separate items into groups
+  absl::flat_hash_map<ValuesList, std::vector<DocValues>> groups;
+  for (auto& value : result.values) {
+    groups[ExtractFieldsValues(value, fields)].push_back(std::move(value));
   }
 
-  absl::FixedArray<Value> Extract(const DocValues& dv) {
-    absl::FixedArray<Value> out(fields_.size());
-    for (size_t i = 0; i < fields_.size(); i++) {
-      auto it = dv.find(fields_[i]);
-      out[i] = (it != dv.end()) ? it->second : Value{};
+  // Restore DocValues and apply reducers
+  auto& values = result.values;
+  values.clear();
+  values.reserve(groups.size());
+  while (!groups.empty()) {
+    auto node = groups.extract(groups.begin());
+    DocValues doc = PackFields(std::move(node.key()), fields);
+    for (auto& reducer : reducers) {
+      doc[reducer.result_field] = reducer.func({reducer.source_field, node.mapped()});
     }
-    return out;
+    values.push_back(std::move(doc));
   }
 
-  DocValues Unpack(absl::FixedArray<Value>&& values) {
-    DCHECK_EQ(values.size(), fields_.size());
-    DocValues out;
-    for (size_t i = 0; i < fields_.size(); i++)
-      out[fields_[i]] = std::move(values[i]);
-    return out;
+  auto& fields_to_print = result.fields_to_print;
+  fields_to_print.clear();
+  fields_to_print.reserve(fields.size() + reducers.size());
+
+  for (auto& field : fields) {
+    fields_to_print.insert(field);
+  }
+  for (auto& reducer : reducers) {
+    fields_to_print.insert(reducer.result_field);
   }
+}
 
-  std::vector<std::string> fields_;
-  std::vector<Reducer> reducers_;
-};
+void Aggregator::DoSort(std::string_view field, bool descending) {
+  auto comparator = [&](const DocValues& l, const DocValues& r) {
+    auto l_it = l.find(field);
+    auto r_it = r.find(field);
 
-const Value kEmptyValue = Value{};
+    // Handle cases where one of the fields is missing
+    if (l_it == l.end() || r_it == r.end()) {
+      return l_it != l.end() || r_it == r.end();
+    }
+    if (l_it->second < r_it->second) {
+      return !descending;
+    }
+    if (l_it->second > r_it->second) {
+      return descending;
+    }
+    return true;  // Elements are equal
+  };
```

Comment:
why do we return true if elements are equal for less operator? As I understand you could take previous implementation and return something like "res != descending"

Reply:
Yes, it makes sense to return `descending` here. But just to note, we are not guaranteeing any strict ordering in the result when elements are equal. This is because the final order also depends on the initial distribution of keys across shards (results from shards are joined into one vector, and then a sort is performed on this array).

So, it is not a "stable" sort, meaning that the initial order of the keys is influenced by the key distribution across shards, which may give the user the impression of an unstable sort